Add ipc router shutdown

gterzian opened the pull request:
<!-- Please describe your changes on the following line: -->

Explicitly shutting down IPC router as part of shutdown, for contexts see https://github.com/servo/servo/pull/25685#pullrequestreview-364604068

Note that the call to `shutdown` is idempotent, so it shouldn't matter whether the call in the constellation and the script-thread are actually on the same `ROUTER` in single-process mode...

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

gterzian committed:
shutting down IPC router thread when a process shuts down.

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/Cargo.toml, components/constellation/Cargo.toml, components/script/script_thread.rs, components/constellation/constellation.rs, components/webdriver_server/Cargo.toml
 * @cbrewster: components/constellation/Cargo.toml, components/constellation/constellation.rs
 * @jgraham: components/webdriver_server/Cargo.toml
 * @kichjang: components/script/Cargo.toml, components/script/script_thread.rs, components/net/Cargo.toml, components/net_traits/Cargo.toml, components/script_traits/Cargo.toml
 * @emilio: components/layout/Cargo.toml

Manishearth reviewed (approved):
looks good so far, unsure if this is still WIP

jdm commented:
```
error: mismatched closing delimiter: `)`
    --> /repo/components/script/script_thread.rs:2887:60
     |
2884 |     fn handle_exit_script_thread_msg(&self) {
     |                                             - unclosed delimiter
...
2887 |         debug!"Exiting IPC router thread in script thread.");
     |                                                            ^ mismatched closing delimiter
```

gterzian commented:
Thanks for reviewing. I've fixed the error and removed the WIP. We could try a run, or if you want to r it already we'll get a run anyway. I think it should be ok but I actually haven't tried to locally or anything. 

I think the constellation and the script-thread would be the two places where you could have different routers if in multi-process, actually it's kind of interesting to realize the script-thread is sort of the "main thread" of a script-process...

CYBAI commented:
I'm just out of curiosity.

Due to seeing that Greg updated all the `ipc-channel` from `"0.14"` to `"0.14.1"`, I'm wondering if it's worth or better to just run `cargo update -p ipc-channel` to update the `ipc-channel` in lockfile to `0.14.1` so that we don't need to update all the toml files 👀?

gterzian commented:
```
error: failed to select a version for the requirement `ipc-channel = "^0.14.1"`
  candidate versions found which didn't match: 0.14.0, 0.13.0, 0.12.2, ...
  location searched: crates.io index
```

Does this mean we need to update the crate for `ipc-channel`?

@CYBAI thanks for thinking about this, I personally don't know.

CYBAI commented:
> Does this mean we need to upload the latest version for `ipc-channel`?

Just checked [ipc-channel on crates.io](https://crates.io/crates/ipc-channel) and I think yes, we will need to publish the latest version 👀
 
> @CYBAI thanks for thinking about this, I personally don't know.

Welcome! and I learned that command from @Eijebong :D

jdm commented:
I have published 0.14.1.

gterzian commented:
@Manishearth r?

Manishearth commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 6ee4dbf has been approved by `Manishearth`

<!-- @bors-servo r=Manishearth 6ee4dbffdfeb6c152e9193c0b71ac088af4c4dee -->

bors-servo commented:
:hourglass: Testing commit 6ee4dbffdfeb6c152e9193c0b71ac088af4c4dee with merge 252485c96d1ea34d5baeaaf6a4f2008b1d52c91f...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/IIfvGS3zTx6HJ-wEAXnq1w)

gterzian commented:
Ok interesting, so it looks like a layout thread receiving a disconnect error on it's channel from the constellation. 

I think what might have happened previoulsy was that an IPC route would keep it alive, so that the problem was hidden.

I think the actual problem might be that the constellation, when shutting down, doesn't wait for all pipelines to shutdown first. I'll try to add some code to that effect and see if it helps.

gterzian commented:
Ah no that's not true, pipelines are always exited before `handle_shutdown`...

gterzian commented:
@bors-servo try=wpt Let's see if shutting-down the router at the end of both constellation and script-thread shutdown helps...

bors-servo commented:
:hourglass: Trying commit 02dca45c30561b23bb72443a007b0e859331dee7 with merge 335845bddaf0ed20c44ef2a0a69485ac4d4890b6...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/KUTHr1VkSQCysfKwoTTSlA)

gterzian commented:
Ok so the test that is failing was added in the light of https://github.com/servo/servo/issues/11546

But now it's the layout thread that is receiving on a closed receiver from the constellation:

```
  │ called `Result::unwrap()` on an `Err` value: RecvError (thread LayoutThread PipelineId { namespace_id: PipelineNamespaceId(2), index: PipelineIndex(6) }, at components/layout_thread/lib.rs:730)
  │ stack backtrace:
  │    0: servo::backtrace::print
  │    1: servo::main::{{closure}}
  │    2: std::panicking::rust_panic_with_hook
  │              at src/libstd/panicking.rs:513
  │    3: rust_begin_unwind
  │              at src/libstd/panicking.rs:417
  │    4: core::panicking::panic_fmt
  │              at src/libcore/panicking.rs:111
  │    5: core::option::expect_none_failed
  │              at src/libcore/option.rs:1211
  │    6: layout_thread::LayoutThread::start
  │    7: profile_traits::mem::ProfilerChan::run_with_memory_reporting
  │    8: std::sys_common::backtrace::__rust_begin_short_backtrace
  │    9: core::ops::function::FnOnce::call_once{{vtable.shim}}
  │   10: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
  │              at /rustc/45ebd5808afd3df7ba842797c0fcd4447ddf30fb/src/liballoc/boxed.rs:1017
  │   11: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
  │              at /rustc/45ebd5808afd3df7ba842797c0fcd4447ddf30fb/src/liballoc/boxed.rs:1017
  │       std::sys_common::thread::start_thread
  │              at src/libstd/sys_common/thread.rs:13
  │       std::sys::unix::thread::Thread::new::thread_start
  │              at src/libstd/sys/unix/thread.rs:80
  │   12: start_thread
  │   13: __clone
  │ [2020-03-30T09:15:03Z ERROR servo] called `Result::unwrap()` on an `Err` value: RecvError
  └ Pipeline failed in hard-fail mode.  Crashing!
```

cc @asajeffrey

gterzian commented:
Ah I think I know what's happening: we should only shut-down the router thread when a script-thread shuts-down when running in multi-process mode, since in single-process the router might still be used by another event-loop(in this case it's the layout-thread related to another script-thread that needs to use it).

gterzian commented:
Ok, looks like it's fixed.

@Manishearth r?

Manishearth commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 8f4ba36 has been approved by `Manishearth`

<!-- @bors-servo r=Manishearth 8f4ba363037899dd436ab3e796c9ee1236e4fc12 -->

gterzian commented:
Oops I think I forgot to fmt it, I guess the build will fail anyway then, no need to minus approve it...

gterzian commented:
@bors-servo r=Manishearth

bors-servo commented:
:pushpin: Commit 53497fe has been approved by `Manishearth`

<!-- @bors-servo r=Manishearth 53497fe87f21eb363f04048ef19f755bb39c940a -->

bors-servo commented:
:hourglass: Testing commit 53497fe87f21eb363f04048ef19f755bb39c940a with merge 35b18d5205bb729b2c4d178b19bc1d771a6d0f09...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/ENwAyLf3R9-KzoPJ1nHgPA)

jdm commented:
@bors-servo retry
* #25513

bors-servo commented:
:hourglass: Testing commit 53497fe87f21eb363f04048ef19f755bb39c940a with merge c8479c8b9032686f78ca2efe1314fb829fac201b...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/QKxIo_46T-mqg-x272fxbA)
Approved by: Manishearth
Pushing c8479c8b9032686f78ca2efe1314fb829fac201b to master...
